src/backend/distributed/commands/multi_copy.c

```diff
@@ -1195,6 +1233,195 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
+	FmgrInfo *columnOutputFunctions = NULL;
+
+	/* allocate column values and nulls arrays */
+	Relation distributedRelation = heap_open(relationId, RowExclusiveLock);
+	TupleDesc tupleDescriptor = RelationGetDescr(distributedRelation);
+	uint32 columnCount = tupleDescriptor->natts;
+	Datum *columnValues = palloc0(columnCount * sizeof(Datum));
+	bool *columnNulls = palloc0(columnCount * sizeof(bool));
+
+	EState *executorState = CreateExecutorState();
+	MemoryContext executorTupleContext = GetPerTupleMemoryContext(executorState);
+	ExprContext *executorExpressionContext = GetPerTupleExprContext(executorState);
+
+	ShardConnections *shardConnections =
+		(ShardConnections *) palloc0(sizeof(ShardConnections));
+
+	/* initialize copy state to read from COPY data source */
+	CopyState copyState = BeginCopyFrom(distributedRelation,
+										copyStatement->filename,
+										copyStatement->is_program,
+										copyStatement->attlist,
+										copyStatement->options);
+
+	CopyOutState copyOutState = (CopyOutState) palloc0(sizeof(CopyOutStateData));
+	copyOutState->binary = true;
+	copyOutState->fe_msgbuf = makeStringInfo();
+	copyOutState->rowcontext = executorTupleContext;
+
+	columnOutputFunctions = ColumnOutputFunctions(tupleDescriptor, copyOutState->binary);
+
+	/* we use a PG_TRY block to close connections on errors (e.g. in NextCopyFrom) */
+	PG_TRY();
+	{
+		uint64 shardMaxSizeInBytes = (int64) ShardMaxSize * 1024L;
+		uint64 copiedDataSizeInBytes = 0;
+		uint64 processedRowCount = 0;
+
+		/* set up callback to identify error line number */
+		ErrorContextCallback errorCallback;
+
+		errorCallback.callback = CopyFromErrorCallback;
+		errorCallback.arg = (void *) copyState;
+		errorCallback.previous = error_context_stack;
+		error_context_stack = &errorCallback;
+
+		while (true)
+		{
+			bool nextRowFound = false;
+			MemoryContext oldContext = NULL;
+			uint64 messageBufferSize = 0;
+
+			ResetPerTupleExprContext(executorState);
+
+			oldContext = MemoryContextSwitchTo(executorTupleContext);
+
+			/* parse a row from the input */
+			nextRowFound = NextCopyFrom(copyState, executorExpressionContext,
+										columnValues, columnNulls, NULL);
+
+			if (!nextRowFound)
+			{
+				MemoryContextSwitchTo(oldContext);
+				break;
+			}
+
+			CHECK_FOR_INTERRUPTS();
+
+			MemoryContextSwitchTo(oldContext);
+
+			if (copiedDataSizeInBytes == 0)
+			{
+				StartCopyToNewShard(copyOutState, shardConnections, relationId,
+									copyStatement);
+			}
+
+			/* replicate row to shard placements */
+			resetStringInfo(copyOutState->fe_msgbuf);
+			AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
+							  copyOutState, columnOutputFunctions);
+			SendCopyDataToAll(copyOutState->fe_msgbuf, shardConnections->connectionList);
+
+			messageBufferSize = copyOutState->fe_msgbuf->len;
+			copiedDataSizeInBytes = copiedDataSizeInBytes + messageBufferSize;
+
+			if (copiedDataSizeInBytes > shardMaxSizeInBytes)
```

Comment:
Maybe also add a comment here saying what we're trying to achieve by doing this.